File: src/2025/day1.ts
export const solve_day_1 = (input: string): number => {
    const lines: string[] = input.split("\n");

    let pointer: number = 50;
    let count: number = 0;

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (!line) continue;

        const indicator = line[0];
        const angle = parseInt(line.slice(1));

        if (indicator == "L") {
            pointer -= angle;
        } else {
            pointer += angle;
        }

        pointer = rem_euclid(pointer);

        if (pointer == 0) {
            count += 1;
        }
    }

    console.log(count);
    return count;
}

export const solve_day_1_part_2 = (input: string): number => {
    const lines: string[] = input.split("\n");

    let pointer: number = 50;
    let count: number = 0;

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (!line) continue;

        const indicator = line[0];
        const angle = parseInt(line.slice(1));

        const start = pointer;
        const end = indicator === "L" ? pointer - angle : pointer + angle;

        count += count_zeros(start, end);

        pointer = rem_euclid(end);
    }

    console.log(count);
    return count;
}

const count_zeros = (start: number, end: number): number => {
    if (end > start) {
        const first_cross = Math.floor(start / 100 + 1) * 100;
        if (first_cross > end) {
            return 0;
        }
        return Math.floor((end - first_cross) / 100) + 1;
    } else if (end < start) {
        const first_cross = start % 100 === 0 ? start - 100 : Math.floor(start / 100) * 100;
        if (first_cross < end) {
            return 0;
        }
        return Math.floor((first_cross - end) / 100) + 1;
    }
    return 0;
}

const rem_euclid = (value: number, limit: number = 100): number => {
    return ((value % limit) + limit) % limit;
}

File: srcd/bin/2025/day1.dart

class Day1_2025 {

  int pointer = 50;
  int count = 0;
  final List<String> lines;

  Day1_2025._({
    required this.lines
  });

  factory Day1_2025(String input) {
    final List<String> lines = input.split("\n");
    return Day1_2025._(lines: lines);
  }

  (String, int) getDirAndAngle(String s) {
      final String dir = s[0];
      final String idx = s.substring(1);
      final int angle = int.parse(idx);
      return (dir, angle);
  }

  int getResult() {
    pointer = 50;
    count = 0;

    for (var line in lines) {
      if(line.isEmpty) continue;
      final (dir, angle) = getDirAndAngle(line);
      final int tmp = dir == "L" ? pointer - angle : pointer + angle;
      pointer = remEuclid(tmp);
      if (pointer == 0)  count++;
    }
    return count;
  }

  int getResult2() {
    pointer = 50;
    count = 0;

    for (var line in lines) {
      if (line.isEmpty) continue;
      final (dir, angle) = getDirAndAngle(line);

      int start = pointer;
      int end = dir == "L" ? pointer - angle : pointer + angle;

      count += countZeros(start, end);
      pointer = remEuclid(end);
    }

    return count;
  }

  int remEuclid (int value, {int limit = 100}) {
    return ((value % limit) + limit) % limit;
  }

  int countZeros(int s, int e) {
    if (e > s) {
      final firstCross = (s ~/ 100 + 1) * 100;
      if (firstCross > e) {
        return 0;
      }
      return (e - firstCross) ~/ 100 + 1;
    } else if (e < s) {
      final firstCross = s % 100 == 0 
          ? s - 100 
          : (s ~/ 100) * 100;
      if (firstCross < e) {
        return 0;
      }
      return (firstCross - e) ~/ 100 + 1;
    }
    return 0;
  }
}